use stat command with 300ms timeout to get device ids for IsLikelyNotMountPoint

manugupt1 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

<!--
Add one of the following kinds:

/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:



syscall.Stat can get blocked on NFS calls; therefore the kubelet
can get wedged as seen in
https://github.com/kubernetes/kubernetes/issues/101622

Using a go routine would lead to a go routine leak and there is
currently no good way to handle the go routine leak.
Ref: https://github.com/golang/go/issues/41054

This change calls stat command to get a device major and minor id
using the stat command within 300ms (because this is supposed to be
a fast operation).

This commit takes in inspiration from this command:
https://github.com/kubernetes/kubernetes/pull/102038/files#r644892910

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #101622

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig node storage
/assign @gnufied 

cc @MadhavJivrajani @pacoxu

jingxu97 commented on staging/src/k8s.io/mount-utils/mount_linux.go:
```diff
@@ -48,6 +48,15 @@ const (
+	// statCommandArgs is the arguments that are given to the stat command. Returns device major, minor number.
+	statCommandArgs = `--format={"major": %t, "minor": %T}`
+	// statTimeout is used to time out the stat command after 300 ms. This is randomly chosen; but
+	// has been kept less than the nfsTimeout.
```
what is the nfsTimeout?
is 300ms small?

manugupt1 replied:
Apologies: 
https://github.com/kubernetes/kubernetes/blob/4ad08fa5b611826505fa2892cdce463f12d32b35/pkg/volume/nfs/nfs.go#L67

I do not think so it is small since `IsLikelyNotMountPoint` is intended to be fast.

wongma7 commented:
/ok-to-test

manugupt1 commented:
Seems like all the e2e tests are failing because `stat` is not installed. This might not work.

Commit:
use stat command with 300ms timeout to get device ids for IsLikelyNotMountPoint

syscall.Stat can get blocked on NFS calls; therefore the kubelet
can get wedged as seen in

Using a go routine would lead to a go routine leak and there is
currently no good way to handle the go routine leak.
Ref: golang/go#41054

This change calls stat command to get a device major and minor id
using the stat command within 300ms (because this is supposed to be
a fast operation).

This commit takes in inspiration from this command:
https://github.com/kubernetes/kubernetes/pull/102038/files#r644892910

ydp commented:
@manugupt1  why close this? see you added the path for stat command not existing scenario, any other concerns?

manugupt1 commented:
@ydp not sure why I closed it; probably because I did not want to add fallback behavior on top of fallback behavior.
